oio/directory/indexer.py

```diff
@@ -0,0 +1,257 @@
+        self.full_scan_nb += 1
+        self.success_nb = 0
+        self.failed_nb = 0
+        now = time.time()
+        self.last_report_time = now
+
+        self.report("starting")
+
+        for db_path in paths:
+
+            # Graceful exit, hopefully
+            if self.stop:
+                break
+
+            db_id = db_path.rsplit("/")[-1]
+            db_id = db_id.rsplit(".")
+
+            if len(db_id) != 3:
+                self.warn("Malformed db file name !", db_path)
+                continue
+
+            db_id = ".".join([db_id[0], db_id[1]])
+            self.index_meta2_database(db_id)
+
+            self.last_index_time = ratelimit(
+                self.last_index_time,
+                self.max_indexed_per_second
+            )
+
+            now = time.time()
+            if now - self.last_report_time >= self.report_interval:
+                self.report("running")
+
+        self.report("ended")
+
+    def run(self):
+        """
+        Main worker loop
+        """
+        self.start_time = time.time()
+        while not self.stop:
+            try:
+                self.crawl_volume()
+                self.last_scan_time = time.time()
+                time.sleep(self.scans_interval)
+            except exc.OioException as e:
+                self.logger.exception("ERROR during indexing meta2: %s", e)
+
+    def stop(self):
+        """
+        Could be needed for eventually gracefully stopping.
+        """
+        self.stop = True
+
+
+def run(worker):
+    """
+    Simple workaround to run the Meta2IndexingWorker in a ContainerPool.
+    :param worker Meta2IndexingWorker The worker to run.
+    """
+    worker.run()
+
+
+class Meta2Indexer(Daemon):
+    """
+    A daemon that spaws a greenlet running a Meta2IndexingWorker
+    for each volume.
+
+    Possible values of conf relating to this worker are:
+        - volume_list:
+
+        - interval: (int) in sec time between two full scans.
+            Default: half an hour.
+        - report_interval: (int) in sec, time between two reports.
+            Default: 300
+        - scanned_per_second: (int) maximum number of indexed databases /s.
+            Default: 1000
+    """
+
+    def __init__(self, conf):
+        super(Meta2Indexer, self).__init__(conf=conf)
+        self.logger = get_logger(conf)
+        if not conf.get("volume_list"):
+            raise exc.OioException("Not meta2 volumes provided to index !")
+        self.volumes = conf.get('volume_list').split(',')
+        if len(self.volumes) == 0:
+            raise exc.OioException("Not meta2 volumes provided to index !")
+        self.volumes = [x.strip() for x in self.volumes]
+        self.pool = ContextPool(len(self.volumes))
+        self.volume_workers = [Meta2IndexingWorker(x, conf) for x in
+                               self.volumes]
+
+    def run(self, *args, **kwargs):
+        for worker in self.volume_workers:
+            self.pool.spawn(run, worker)
```

Comment:
I think it's possible to pass `Meta2IndexingWorker.run` instead of the "workaround" method.